### PR TITLE
Add a single source of build and release configuration

### PR DESCRIPTION
**Pull request by maxsokolovsky**
## Problem

Release captains need to painstakingly edit the values of many things before the release in different files, like Dockerfiles, Go files, etc. This is inconvenient and error-prone.
 
## Solution
This PR introduces a single manifest with all config values.
In the future, this will be the single place for a release captain to update our configuration values for releases and pre-releases.

As our Dockerfiles and Go files aren't structured data, we don't want to do any text manipulation with `sed` to replace text. Instead, we do the following:
- If we're building Docker images, craft the `docker build` command to use build-args with values replaced with those in the manifest.
- If we're updating some string literal in a Go file, we refer to a global variable from a special package generated on any config change.

This pilot PR migrates a few values:
1.  rancher-webhook version for Rancher and agent
1. default shell image version
1. CSP adapter version
1. Fleet version 

The other values can be migrated later.
Note that there is no need to specify them in any Dockerfiles. Use the `build.yaml` manifest instead.

See the new docs in `/docs/build.md`.
 
### Manual Testing
Specified a value and build first the main Rancher, then the agent. Pushed up both to DockerHub and created a Rancher instance from the image, then made a downstream cluster that deployed the custom agent. Observed that the versions of the webhook in the local and downstream clusters are the same and match the one specified in the Dockerfile in one place.

### Automated Testing
None - Lack of the framework capable of testing this fix/change

### Regressions Considerations

Existing / newly added automated tests that provide evidence there are no regressions:

### CODE REVIEW

File: package/Dockerfile
Comment (cbron) [inline package/Dockerfile:236]: Relying on an actual file feels weird. Perhaps instead of fighting just this one env var we make a script system that is the bases for some release automation that runs locally pre-build and puts the right versions (of whatever) in the right files through some `sed` magic. What you think @MbolotSuse , we've discussed this before.

Comment (maxsokolovsky) [reply package/Dockerfile]: I've redone the solution.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Two Overall Comments:
- There's also a script that is used for building dev images here: https://github.com/rancher/rancher/blob/release/v2.7/dev-scripts/build-local.sh. Please update that as well (optional but would be nice).
- Please add some docs somewhere on how this works (i.e something in the readme explaining what the build file is and how that gets into the code base).

File: Dockerfile.dapper
Comment (MbolotSuse) [inline Dockerfile.dapper:34]: Nit:

If you can list the YQ version as an env var above (like we do for helm unit test/helm/kustomize) that would be good. It makes it easier for our versioning automation to pick up and do automatic bumps of. If you have a second I would appreciate you doing that for the controller-gen utility in the line before this one as well, I missed that during that PR review.

File: scripts/ci
Comment (MbolotSuse) [reply scripts/ci:7]: Now that you have moved this into `package` directly, I don't think we need it here.

File: release.md
Comment (MbolotSuse) [inline release.md:4]:
```suggestion
Rancher relies on various sub-components, such as the webhook. These typically need to have set versions for Rancher to build properly.
```

File: release.md
Comment (MbolotSuse) [inline release.md:5]:
```suggestion
Build variables can be used in a variety of places/ways, including as Env Vars in Dockerfiles, and Constants in go code. 
```

File: release.md
Comment (MbolotSuse) [inline release.md:13]:
```suggestion
Next, re-build Rancher using `make build`, or re-package rancher using `make package`.
```

File: release.md
Comment (MbolotSuse) [inline release.md:18]: Would also be good to include a general guide on updating the places that it is used - the below will create a new build var but it won't actually get used when building rancher.

**Review by MbolotSuse — COMMENTED** [review]
Couple of comments on the new `release.md` file

File: scripts/config.yaml
Comment (MbolotSuse) [inline scripts/config.yaml:1]: I would recommend moving this up one directory into the root of the repo if possible. This file contains all of the versions of the sub-applications, it should not be hidden in scripts if we can avoid it.

File: release.md
Comment (superseb) [inline release.md:1]: Please edit or delete `docs/release.md` if this will be the new location.

Comment (maxsokolovsky) [reply release.md:1]: I think we could delete it anyways, since we haven't been using it, as far as I understand. Release captains use instructions found in Confluence...

File: scripts/package
Comment (MbolotSuse) [inline scripts/package:4]: Same comment about needing `$(dirname $0)/config`

File: scripts/build
Comment (MbolotSuse) [reply scripts/build:4]: Nit: This should probably be at the lowest level possible for usage (i.e. `build-server` and `build-agent`) - my concern here is that if someone calls `build-server` that should work on it's own, but it won't in this current form.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]

File: scripts/config
Comment (maxsokolovsky) [reply scripts/config:11]: You're right, basename is not needed here anymore.

**Review by MbolotSuse — APPROVED** [review]
LGTM, though you should get reviewers from other teams so that everyone gets a chance to look at the new configuration file.

File: README.md
Comment (cbron) [inline README.md:34]: Why'd all this change, is this unrelated ? Seems like there was a max line setting somewhere. I would prefer to revert all this, as it is less readable IMO.

File: config.yaml
Comment (cbron) [inline config.yaml:1]: My biggest callout here is to make sure the release checklist gets updated at some point. At least with anything that affects the wider build. For now we own webhook, but once this moves to charts or KDM, the checklist needs to point here.

File: docs/build.md
Comment (cbron) [inline docs/build.md:25]: Is there an example for go one ? I don't see in this PR. Overall this is interesting approach, and clever. I'm not sure it will scale well to other parts of the release process. For example we can't use env-vars to change things in setting.go, as env-vars are already treating as a second-class citizen there. I suppose the ldflags part will work, but that could get really messy if there are 50 ldflags, and taking the value out of setting.go will be confusing.  But for now thats ok, we can refactor once we get further along. 

> As as our Dockerfiles and Go files aren't structured data, we don't want to do any text manipulation with sed to replace text

The benefit to using `sed` or any other text manipulation tool is that you are guaranteed to not having problems at build or run time, everything is left in the developers hands and gets code-reviewed on the way in. I still want to consider that as an option to see how well this scales.

**Review by cbron — COMMENTED** [review]
Minor comments then lgtm

File: Dockerfile.dapper
Comment (cbron) [inline Dockerfile.dapper:29]: On testing, when there is a downstream cluster back in the integration tests, it would be good to check if that version is what is expected.

File: README.md
Comment (superseb) [inline README.md:1]: `*This file is auto-generated from README-template.md, please make any changes there.*` 

This needs to be changed in `README-template.md`

File: README.md
Comment (maxsokolovsky) [reply README.md:34]: I needed to make a small update to the file. When I saved it, Goland ran the default Markdown formatter, and I chose to keep the changes. Right, the max line is there and line breaks between headers and text, which I think is good, no? It's easier to read raw Markdown, since you don't need to scroll to the right all the time, but it's rendered the same...

File: docs/build.md
Comment (maxsokolovsky) [reply docs/build.md:25]: No example of a Go variable yet, since we initially just wanted to port over the webhook version, which is an env var. But I can add one for something, like the shell version, which would be a variable.

File: README.md
Comment (maxsokolovsky) [reply README.md:1]: I see! So I shouldn't change `README.me` at all in this PR? Instead, I could just change the template and run the workflow manually after merging this?

File: docs/build.md
Comment (MbolotSuse) [reply docs/build.md:25]: > The benefit to using sed or any other text manipulation tool is that you are guaranteed to not having problems at build or run time, everything is left in the developers hands and gets code-reviewed on the way in. I still want to consider that as an option to see how well this scales.

The only requirement in the current format is that your target value goes in the `config.yaml` - everything else is left up to how that specific value is persisted in Rancher (wherever that may be). This is all to say that sed is still an option for our build scripts.

That being said, I very much dislike sed based pipelines. I would disagree that they guarantee not having problems at build or run time - they can result in un-intended variables being changed (i.e. when replacing `CATTLE_RANCHER_WEBHOOK_VERSION` did you also update `CATTLE_RANCHER_WEBHOOK_MIN_VERSION` by accident?). I also think they are much worse in terms of read-ability.

But we can evaluate this as time goes by, we don't need to make a decision today.

Comment (cbron) [reply docs/build.md:25]: Agree, we can solve later. I'm going to leave this thread unresolved.

File: README.md
Comment (cbron) [reply README.md:34]: I suppose since it renders the same it doesn't matter. IMO readability is hit and miss, and better left up to the writer. For example the _"please refer to our Rancher Documentation"_ sentence is now on two lines instead of one. Same for "Please see installation requirements" or "open source libraries and projects". Could you just increase the line length ? If not I would prefer to revert.

File: scripts/chart/build
Comment (KevinJoiner) [inline scripts/chart/build:19]: this should pull in the environmental variable from sourcing export-config.

File: scripts/build-agent
Comment (KevinJoiner) [inline scripts/build-agent:5]: Why are we sourcing here if this file does not use any variables?

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: scripts/export-config
Comment (maxsokolovsky) [reply scripts/export-config:15]: I did it like this originally, but shellcheck issued warnings. We can ignore them and just have a single statement for declaration and export instead, like in your suggestion.

File: scripts/chart/build
Comment (maxsokolovsky) [reply scripts/chart/build:19]: Sure, that works better. I had originally just kept this Go source file parsing behavior.

File: tests/v2/codecoverage/scripts/build_docker_images.sh
Comment (maxsokolovsky) [reply tests/v2/codecoverage/scripts/build_docker_images.sh:33]: This deviation is deliberate. Could you clarify why we need the braces here? We don't plan to append anything to the expanded variable, so those two bits of syntax are equivalent for our purposes. But without braces, it's simpler, that's why I chose to make my variables look like this.

File: pkg/buildconfig/writer.go
Comment (maxsokolovsky) [reply pkg/buildconfig/writer.go:1]: I am not sure I agree. This package provides a type that could be re-used elsewhere in our packages.
What do you mean by a package writing its own files?

File: pkg/buildconfig/writer.go
Comment (maxsokolovsky) [reply pkg/buildconfig/writer.go:87]: But constants would need to have concrete types (numbers and floats). Do you suggest I simply remove any quotes (if any are present for numbers) when I render the template? That could work...

I should totally add support to bools regardless (good catch). I am not sure how machine-dependent size of ints is relevant here, though.

File: pkg/buildconfig/writer.go
Comment (maxsokolovsky) [reply pkg/buildconfig/writer.go:33]: The nil checks are for internal needs. I wanted to have a simpler interface to this, which is why there is only one exported function on the type. And I do say in the comment that each instance of the struct is to be used only once.

File: README-template.md
Comment (maxsokolovsky) [reply README-template.md:9]: This was caused by the standard Markdown formatter - it made it tidy without changing how the text is rendered.

File: pkg/buildconfig/writer.go
Comment (maxsokolovsky) [reply pkg/buildconfig/writer.go:61]: I am inclined to agree here. I just thought it's a special case of error that we want to bring to someone's attention - since each value of this type is to be used only once.

File: scripts/build-agent
Comment (maxsokolovsky) [reply scripts/build-agent:5]: I might have added this here unnecessarily, yeah. Although for the future, this wouldn't hurt - if any variables will be used in the agent binary build...
Will remove if you think it's best.

File: README-template.md
Comment (KevinJoiner) [inline README-template.md:32]: Remove IDE reformats.

File: scripts/chart/build
Comment (KevinJoiner) [reply scripts/chart/build:19]: We should use the shell environmental varible since this use case is part of why we introduced the change in the first place.

File: tests/v2/codecoverage/scripts/build_docker_images.sh
Comment (KevinJoiner) [inline tests/v2/codecoverage/scripts/build_docker_images.sh:6]: We don't need to call cd here
```suggestion
source $(dirname $0)/../../../../scripts/export-config
```

File: pkg/codegen/buildconfig/writer.go
Comment (KevinJoiner) [inline pkg/codegen/buildconfig/writer.go:112]: nit: 
Add package comment for our generated file.

File: docs/build.md
Comment (KevinJoiner) [inline docs/build.md:22]: The current wording sounds like the user needs to perform steps 1 or 2.
We should update the wording to explain this is what the build scripts are doing.
```suggestion
The build configuration passes a configuration value in one of two ways:
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: docs/build.md
Comment (maxsokolovsky) [reply docs/build.md:22]: No, this is indeed for devs who want to add a new value to the build-release system. This is a one-time operation. For example, you want to add support for the charts branch - you follow these instructions.

File: docs/build.md
Comment (maxsokolovsky) [reply docs/build.md:25]: I'm sorry, perhaps I'm not understanding your suggestion...
We set `build-args` when we run the `docker build` command, which is common knowledge and also documented in my example on line 25. In this case, it's the scripts that run the `build` command, not even the devs themselves.

File: docs/build.md
Comment (KevinJoiner) [reply docs/build.md:22]: > you follow these instructions.

Your numbered points are not instructions they are explaining what the build configuration is doing. 

`Pass a configuration value in one of two ways:` Is implying that we are going to tell the user how to pass the configuration value. But the first step says `This is not something the dev is doing.`

I understand what you are trying to say because I am familiar with the process but for a new dev they would expect the numbered sections to be steps to perform instead you are using them as section headers.

> 1. As a `--build-arg` to `docker image build`. Capture it as an `ARG` in the Dockerfile
   via `docker build --build-arg MYVAR="$MYVAR" ...`
 As a reference to a constant found in the generated Go [file](../pkg/buildconfig/constants.go).
This is not something the dev is doing. 
> 2. As a reference to a constant found in the generated Go [file](../pkg/buildconfig/constants.go).

Comment (KevinJoiner) [reply docs/build.md:22]: 👀

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

**Review by KevinJoiner — DISMISSED** [review]

File: pkg/codegen/buildconfig/writer_test.go
Comment (JonCrowther) [inline pkg/codegen/buildconfig/writer_test.go:1]: Nit: Maybe add a test to ensure the contents get sorted, or even just switch the order of `contents` in `TestGoContantsWriterRun`

**Review by JonCrowther — APPROVED** [review]
LGTM, one suggestion for an additional test

File: Dockerfile.dapper
Comment (aiyengar2) [inline Dockerfile.dapper:36]: If we are making this change, should we move all the `ENV` variables to the top to keep it consistent?

File: build.yaml
Comment (aiyengar2) [inline build.yaml:4]: Can we have the keys in this YAML file directly correspond to environment variable names? That way we don't need one more map of what value here corresponds to a value in the Dockerfile.

Comment (aiyengar2) [reply build.yaml:4]: Also, why does this have to be a YAML file? Can't we just make this a file like:

```
WEBHOOK_VERSION=103.0.0+up0.4.0
CSP_ADAPTER_VERSION=103.0.0+up3.0.0
...
```

And just directly source this in the script execution?

File: dev-scripts/build-local.sh
Comment (aiyengar2) [inline dev-scripts/build-local.sh:60]: Doesn't this mean we now need to add `--build-arg` whenever we add a field to the `build.yaml`? We should script this to infer the `--build-arg` additions that need to be added by reading the `build.yaml` source (or `build.sh`, if we use the suggestion I have in my review above).

File: pkg/buildconfig/constants.go
Comment (aiyengar2) [inline pkg/buildconfig/constants.go:10]: This doesn't have to be generated via a template. We should leave all of these constants as empty strings and have the value be passed in as a build arg: https://blog.alexellis.io/inject-build-time-vars-golang/

This is how version is injected, see https://github.com/rancher/rancher/blob/58bfcb19605d51bf24f0aa5bf1d60a68b1ae2a27/pkg/version/version.go#L12-L15 and https://github.com/rancher/rancher/blob/58bfcb19605d51bf24f0aa5bf1d60a68b1ae2a27/scripts/build-server#L29-L30

Comment (aiyengar2) [reply pkg/buildconfig/constants.go:10]: If we take this approach, I believe we can entirely eliminate the `go generate` and `buildconfig` logic in favor of just a change to the build script.

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
Overall, I'm fine with the idea of consolidating our constants to one location. However, I think we need to change some things in this PR.

Namely,
1. We should do build-time injection instead of generating code. We can add an init function or something else to verify that all of these variables are set before we start execution with a valid error statement if not.
2. We should consider what the right format for specifying values to the `build.yaml` should be. I'm not sure if YAML is the best approach here, since I would prefer something like a `build.sh` that sets environment variables in a flat list instead. Settings environment variables should be exactly as we expect them to be overridden by users, i.e. `CATTLE_RANCHER_WEBHOOK_VERSION` should represent `RancherWebhookVersion` in settings and the default name in the Dockerfile.

File: build.yaml
Comment (aiyengar2) [reply build.yaml:4]: If we plan to stick with YAML, can we also move other things like https://github.com/rancher/rancher/blob/58bfcb19605d51bf24f0aa5bf1d60a68b1ae2a27/pkg/image/origins.go#L21-L252 into this file?

It would be nice if we had a format like:

```yaml
env:
  WEBHOOK_VERSION: 103.0.0+up0.4.0
  ...
image-origins:
  aci-containers-controller: https://github.com/noironetworks/aci-containers
  ...
```

Comment (maxsokolovsky) [reply build.yaml:4]: 1. Agreed on making the keys correspond to environment variable names, will adjust.
2. It doesn't need to be YAML; YAML is just a decent, natural option. And, like your nested example shows, picking YAML is good for the future, should we decide to have more a complicated structure of keys.
3. As for the `origins.go` file, we can do this, but perhaps in a separate PR. This PR aims to introduce the new system and migrate only a handful of variables (listed in PR description).

File: pkg/buildconfig/constants.go
Comment (maxsokolovsky) [reply pkg/buildconfig/constants.go:10]: When we met within the team to discuss this design, we did have an idea of a build-time variable injection. But we decided against it. I don't recall all the reasons (almost 6 months have passed). Perhaps @MbolotSuse or @KevinJoiner could chime in here.

One issue was that devs may not necessarily be building Rancher using the build script, they just use `go build`. Without explicit injection, the variables will remain initialized with placeholder values, which may produce invalid behavior at runtime.

File: dev-scripts/build-local.sh
Comment (maxsokolovsky) [reply dev-scripts/build-local.sh:60]: Not all values are fit for the Dockerfile. We are selective about which go there and which end up referenced as constants in regular Go code. Thus I'd rather not assume that the Dockerfile should include **all** values as ENV values.

File: Dockerfile.dapper
Comment (maxsokolovsky) [reply Dockerfile.dapper:36]: Sure, that might be good. I only extracted this as an ENV not because it's to be configurable, but to allow dependency bump tools notify us that there is a newer version.
Looks like the Renovate bot does inspect .dapper files - https://github.com/rancher/rancher/pull/41453. This one is just inline curl fetch of a binary.
Not sure why it hasn't notified us of a Kustomize or this controller-gen bumps, for example. Perhaps because they are specified in the ENV directives, actually, rather than inline. Will investigate.

**Review by aiyengar2 — DISMISSED** [review]
approving to unblock merging this PR since my review does not need to be blocking. This action does not imply approval.

File: pkg/buildconfig/constants.go
Comment (aiyengar2) [reply pkg/buildconfig/constants.go:10]: FWIW, I still think build-time injection is a better approach here since go generating a constant file seems too heavyweight to me and it matches the pattern we use to build the docker image using build args.

File: dev-scripts/build-local.sh
Comment (aiyengar2) [reply dev-scripts/build-local.sh:60]: We should document this.

File: build.yaml
Comment (aiyengar2) [reply build.yaml:4]: If we plan to introduce a more complicated structure of keys, I think the keys we introduce here should not be top-level keys.

File: Dockerfile.dapper
Comment (olblak) [reply Dockerfile.dapper:36]: For info Renovate inspect .dapper because there is a rule defined [here](https://github.com/rancher/renovate-config/blob/0596cc041373a388718f6364b9ae12e2a82a635a/default.json#L89)

File: pkg/buildconfig/constants.go
Comment (weyfonk) [inline pkg/buildconfig/constants.go:6]: At the risk of stating the obvious here, we will want to be careful not to accidentally downgrade anything when fixing conflicts to merge this PR.
For instance, Fleet `0.9.0` has been un-RCed.

**Review by weyfonk — APPROVED** [review]
LGTM, with similar [doubts](https://github.com/rancher/rancher/pull/42000#discussion_r1399756906) to @aiyengar2's about build-time injection vs code generation.

Plus a warning about conflict resolution, since some charts have been updated, un-RCed, etc since this PR was first submitted... and Fleet now uses a new `CATTLE_FLEET_VERSION` config variable in addition to the already existing (and now deprecated) min version variable.

File: pkg/buildconfig/constants.go
Comment (maxsokolovsky) [reply pkg/buildconfig/constants.go:6]: Yup, this is normal. I keep fixing conflicts while the PR is open, which is OK. We plan to merge it shortly after 2.8.0 is out. We won't need to migrate all other configurable values to the new system immediately. We can do so gradually - one value at a time.

File: dev-scripts/build-local.sh
Comment (maxsokolovsky) [reply dev-scripts/build-local.sh:60]: We can't document the intended use of each value in advance because we can't know who might want to use one and where. The good thing is that the generated constants file has all of them, whether one plans to use some variable in Go code or not.

File: Dockerfile.dapper
Comment (maxsokolovsky) [reply Dockerfile.dapper:36]: Thanks! Will add a rule once we merge this PR.

**Review by weyfonk — APPROVED** [review]

File: pkg/buildconfig/constants.go
Comment (KevinJoiner) [reply pkg/buildconfig/constants.go]: I don't think we should add build-time injection for required variables since it complicates the usage of a lot of native go commands such as `go run`, `go test`, `go install` and `go build`. build-time injection is primarily used for things like commit/version tags so you know when the build was created, but these values are required at runtime for rancher to work.

On the other hand, we already use `go generate` to create constant files so we are not breaking any patterns. 
I also would not consider `go generate` heavy. I think there may be bias towards Rancher's Controller generation which is heavy, but that is a Rancher problem not a `go generate` problem.

**Review by KevinJoiner — APPROVED** [review]

**Comment by mattfarina** [comment]
cc @manno @thardeck

**Review by weyfonk — APPROVED** [review]

**Review by KevinJoiner — APPROVED** [review]

**Pull request merged**
